### PR TITLE
Shelly dimmer

### LINKED ISSUE #949: Shelly dimmer 2 support
**Describe the problem you have/What new integration you would like**

Add support for the Shelly Dimmer 2. This is an esp8266-based dimmer that is very easy to flash (like the Shelly 1 switches, which work brillantly with ESPHome firmware!). 

In the Dimmer 2, the ESP talks to an STM32 microcontroller on the board that performs the actual dimmers. Others have already reverse-engineered the communications protocol; see e.g. [here](https://github.com/Mollayo/Shelly-Dimmer-2/blob/master/light.cpp). More discussion (involving the Shelly Dimmer *1*) can be found over at the Tasmota project, [here](https://github.com/arendst/Tasmota/issues/6914).

**Please describe your use case for this integration and alternatives you've tried:**

I would like to be able to use the Dimmer 2 like I use the Shelly Switches in ESPHome (e.g. have it function as a light in Home-assistant, for instance).

At this point only the Shelly factory firmware is usable (which works fine but does not integrate with ESPHome, which I use for almost all other equipment). Tasmota support appears to be underway. 

**Additional context**

Thank you very much for making ESPHome! Device management is really convenient and an underestimated aspect of home automation.

### PR DESCRIPTION
**Pull request by jbergler**
# What does this implement/fix? 

This change adds support for the Shelly Dimmer and Shelly Dimmer 2.

Very little of this code is my own, most of the work is the [contribution](https://github.com/esphome/feature-requests/issues/949#issuecomment-751696428) of @kostko.
I've simply taken the changes from the branch they shared and separated them into a few smaller, logically separate commits to hopefully make reviewing easier.
From there I made some minor changes to support the ColorMode changes there were recently merged.

The firmware for the co-processor is from this project: https://github.com/jamesturton/shelly-dimmer-stm32
Which I think has primarily been used in tasmota.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Feature request:** fixes https://github.com/esphome/feature-requests/issues/949

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#? <- will provide this shortly.

## Test Environment

- [ ] ESP32
- [X] ESP8266

## Example entry for `config.yaml`:

```yaml

# disable serial, we need this to talk to the co-processor
logger:
  baud_rate: 0

# define a light 
light:
  - platform: shelly_dimmer
    id: dimmer
    name: Dimmer
    leading_edge: false
    min_brightness: 0
    max_brightness: 1000
    gamma_correct: 1
    default_transition_length: 1s
    restore_mode: RESTORE_DEFAULT_OFF
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by jbergler** [comment]
This doesn't need review yet, I'll clean up all the style issues, write docs, and whatever else first.

**Comment by Bascht74** [comment]
@jbergler  Hi, is there a way to test your new component while you do the clean up?

e.G. via: 

```
external_components:
  - source:
      type: git
      url: ...
      ref: ...
    components: [ shelly_dimmer ]
```

I think this will help some people to get it to work with ESP 2021.8.x Releases (like me :-) ...

Thx for a hint...

Sebastian

Update: 
```
external_components:
  - source:
      type: git
      url: https://github.com/jbergler/esphome.src
      ref: shelly-dimmer
    components: [ shelly_dimmer ]
```

I hope this is right / OK.

File: esphome/components/shelly_dimmer/shelly_dimmer.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
+#include <HardwareSerial.h>
+
+#include "esphome/core/component.h"
+#include "esphome/core/esphal.h"
```
Comment (playya) [inline esphome/components/shelly_dimmer/shelly_dimmer.h:4]: This line needs to be replaced by esphome/core/hal.h once you rebase your changes to dev.

**Comment by playya** [comment]
I tested this code using a Shelly Dimmer 2 using figure 1 inhttps://shop.allnet.de/media/pdf/ce/35/07/187930-Handbuch_EN.pdf
I only see flickering if I update STM firmware to 51.6.

**Comment by mafrosis** [comment]
@playya The flickering is not present in v51.5?

**Comment by playya** [comment]
@mafrosis I use Shelly DImmer 2 and I cannot see the flickering.
I update the firmware to 51.6 and had some flickering.

I pushed a branch which includes the firmware update if someone wants to test: https://github.com/playya/esphome/tree/playya/feature/shelly-dimmer-2-fw-51.6
Branch with firmware 51.5: https://github.com/playya/esphome/tree/playya/feature%2Fshelly-dimmer-2

**Comment by jasperslot** [comment]
I would love to see the Shelly dimmer 2 supported! What is necessary to get this moving forward? Can I try this already by using the config example and using this:
```
external_components: - source: type: git url: https://github.com/jbergler/esphome.src ref: shelly-dimmer components: [ shelly_dimmer ]
```
Or do I miss something?

**Comment by playya** [comment]
I started to update the documentaion repository and created 2 yaml files for shelly dimmers available at https://github.com/playya/esphome-docs/tree/shelly_dimmers/devices
Actually 2 pins on ESP are swapped for dimmer2.

On top of those changes I'm using the following to pull my repository:

> external_components:
> 
>   - source:
>       type: git
>       url: https://github.com/playya/esphome
>       ref: playya/feature/shelly-dimmer-2
>     refresh: 1seconds
>     components: [ shelly_dimmer ]
>

**Comment by kentora** [comment]
From the yaml configuration, I couldn't quite figure out how to specify which switch to use as input to the dimming and how.
Is it handled internally?

**Comment by playya** [comment]
To use the Dimmer Code you need to configure the Platform:
https://github.com/playya/esphome-docs/blob/de7336793d35918b2500181361f51cecfda075fd/devices/shelly_dimmer2.yaml#L70

I'm using the Dimmer From Home Assistent, nur If you want to Control it
internally you can Check a more complex example Here:
https://github.com/esphome/feature-requests/issues/949#issuecomment-877807666

kentora ***@***.***> schrieb am Sa., 11. Dez. 2021, 09:55:

**Pull request closed without merging**